private/bufpkg/bufimage/util.go

```diff
@@ -118,11 +203,48 @@ func imageWithOnlyPaths(image Image, fileOrDirPaths []string, allowNotExist bool
 				return nil, fmt.Errorf("path %q has no matching file in the image", potentialDirPath)
 			}
 		}
+		for excludeFileOrDirPath := range excludeFileOrDirPathMap {
+			if _, ok := matchingPotentialExcludePathMap[excludeFileOrDirPath]; !ok {
+				// no match, this is an error given that allowNotExist is false
+				return nil, fmt.Errorf("path %q has no matching file in the image", excludeFileOrDirPath)
+			}
+		}
 	}
 	// we finally have all files that match fileOrDirPath that we can find, make the image
 	return getImageWithImports(image, nonImportPaths, nonImportImageFiles)
 }
 
+func excludeFile(
+	fileMatchingPathMap map[string]struct{},
+	fileMatchingExcludePathMap map[string]struct{},
+) bool {
+	if len(fileMatchingPathMap) == 0 {
+		return true
+	}
+	if len(fileMatchingExcludePathMap) == 0 {
+		return false
+	}
+	seenExcludePaths := make(map[string]struct{})
+	for fileMatchingPath := range fileMatchingPathMap {
+		for fileMatchingExcludePath := range fileMatchingExcludePathMap {
+			if normalpath.EqualsOrContainsPath(fileMatchingPath, fileMatchingExcludePath, normalpath.Relative) {
+				delete(fileMatchingPathMap, fileMatchingPath)
+				seenExcludePaths[fileMatchingExcludePath] = struct{}{}
+				continue
+			}
+		}
+	}
+	// For paths that have never been seen for excludes, we should trim them, since they are
+	// outside of the target paths found.
+	for fileMatchingExcludePath := range fileMatchingExcludePathMap {
+		if _, ok := seenExcludePaths[fileMatchingExcludePath]; ok {
+			continue
+		}
+		delete(fileMatchingExcludePathMap, fileMatchingExcludePath)
+	}
+	return excludeFile(fileMatchingPathMap, fileMatchingExcludePathMap)
```

Comment:
I think the exclusion process you have above makes sense, but I don't think the recursive call buys us anything. From what I can tell, this function would _always_ conclude in a couple recursive calls:

1. All of the relevant files in `fileMatchingPathMap` are removed according to `fileMatchingExcludePathMap`, then `fileMatchingExcludePathMap` shrinks (to approach the base case) according to which paths weren't matched.
2. All of the files that would have been matched by `fileMatchingExcludePathMap` have already been matched (according to (1)). We don't mark any of the `fileMatchingExcludePathMap` as seen, so _all_ of them get deleted.
3. Now, `len(fileMatchingExcludePathMap) == 0`, so we exit via the base case.

In short, I think we can simplify all of this into (1).

Reply:
We had a quick chat during stand-up about this, going to include better context for this in a comment, as mentioned above.